File: Fortran/duffing_ab3.f90
!==================================================================
!==============      DUFFING DAMPED OSCILLATOR      ===============   
!==================================================================


PROGRAM duffing_oscillator

	IMPLICIT NONE
	
	REAL :: ti, tf, h, temp
	INTEGER, PARAMETER :: N=600000
	REAL, DIMENSION(N,2) :: sol
	REAL, DIMENSION(2) :: x0
	REAL :: omega=1

	ti=0
	tf=40000

	h=(tf-ti)/N
	
	x0(1) = 0
	x0(2) = 0
	
	CALL SYSTEM('clear')
	WRITE(*,*) ""
	WRITE(*,*) "======================================================="
	WRITE(*,*) "===========    DUFFING DAMPED OSCILLATOR    ==========="
	WRITE(*,*) "===========       AB3 EXPLICIT SCHEME       ==========="
	WRITE(*,*) "======================================================="
	WRITE(*,*) " ..        . "
	WRITE(*,*) " X + delta x + alpha x + beta x^3 = gamma cos(omega t) "
	WRITE(*,*) ""
	WRITE(*,*) "======================================================="
	
	
	WRITE(*,*) "Time interval:"
	WRITE(*,'(F12.2,F12.2)') ti,tf
	WRITE(*,*) "Number of time steps:"
	WRITE(*,*) N
	WRITE(*,*) "Initial conditions:"
	WRITE(*,'(F10.3,f10.3)') x0(1), x0(2)
	WRITE(*,*) ""
	
	CALL adams_bashforth(sol,h,x0,N)
	
	CALL poincare_map(sol,N,h,omega)
	
	WRITE(*,*) "-------------------------------------------------------"
	WRITE(*,*) "Solution WRITTEN on 'duffing_ab3.dat'"
	WRITE(*,*) "Poincarè Map WRITTEN on 'duffing_ab3_pm.dat'"
	
	


CONTAINS


!==================================================================
!	ADAMS-BASHFORTH 3 EXPLICIT SCHEME IMPLEMENTATION
	SUBROUTINE adams_bashforth(sol,h,x0,N)
	
		REAL, DIMENSION(N,2) :: sol
		REAL, DIMENSION(2) :: x0
		INTEGER :: N, i
		REAL :: h
		
		! INITIALIZE SOLUTION WITH
		
		OPEN(1, file='duffing_ab3.dat')
		
		sol(1,:)=x0
		sol(2,:)=sol(1,:)+h*duffing(h, sol(1,:))
		sol(3,:)=sol(2,:)+h*duffing(2*h, sol(2,:))
		
		WRITE(1,*) 0, sol(1,:)	
		WRITE(1,*) 1*h, sol(2,:)	
		WRITE(1,*) 2*h, sol(3,:)	
		
		DO i=3,(N-1)
			
			sol(i+1,:)=sol(i,:)+h/12*(23*duffing(i*h,sol(i,:)) &
			& -16*duffing((i-1)*h,sol(i-1,:))+5*duffing((i-2)*h,sol(i-2,:)))
			
			WRITE(1,*) (i*h), sol(i+1,:)	

		END DO
		CLOSE(1)
	
	
	END SUBROUTINE adams_bashforth


!==================================================================
!	CONSTRUCTION OF THE POINCARE MAP BY COMPUTED SOLUTION
	SUBROUTINE poincare_map(sol,N,h,omega)
	
		IMPLICIT NONE 
		
		REAL, DIMENSION(N,2) :: sol
		INTEGER :: N, i, ind
		REAL :: h, omega
		REAL :: pi=4*atan(1.d0)
		
		! INITIALIZE SOLUTION WITH
		
		OPEN(2, file='duffing_ab3_pm.dat')
		
		DO i=1,N
			ind=CEILING(1/h*(2*pi*i/omega))
			IF (ind.LE.N) THEN
				WRITE(2,*) i*h, sol(ind,:)
			END IF
		END DO
		
		CLOSE(2)
	
	
	END SUBROUTINE poincare_map


!==================================================================
!	IMPLEMENTING DUFFING OSCILLATOR I ORDER EQUIVALENT SYSTEM	
	FUNCTION duffing(t,x)
		
		REAL, DIMENSION(2) :: duffing
		REAL, DIMENSION(2) :: x
		REAL :: delta=0.15, alpha=-1, beta=1, gam=0.3, omega=1
		REAL :: t


		duffing(1)=x(2)
		duffing(2)=-delta*x(2)-alpha*x(1)-beta*x(1)**3+gam*cos(omega*t)

	END FUNCTION duffing





END PROGRAM duffing_oscillator

File: Python/duffing_ab3.py
from cmath import pi
from math import ceil
import numpy as np
import matplotlib.pyplot as plt

# ==============================================================================
#  DEFINE DYNAMICAL SYSTEM
def Duffing(t,x, params):
    alpha = params["alpha"]
    beta = params["beta"]
    gamma = params["gamma"]
    delta = params["delta"]
    omega = params["omega"]

    xdot = np.array([x[1], -delta*x[1]-alpha*x[0]-beta*x[0]**3+gamma*np.cos(omega*t)])

    return xdot



# ==============================================================================
#  RK4 EXPLICIT SCHEME
def RungeKutta4(f, x0, t0, tf, dt):

    t = np.arange(t0, tf, dt)
    nt = t.size

    nx = x0.size

    x=np.zeros((nx,nt))

    x[:,0] = x0

    for k in range(nt -1):
        k1 = dt*f(t[k], x[:,k])
        k2 = dt*f(t[k] + dt/2, x[:,k]+ k1/2)
        k3 = dt*f(t[k] + dt/2, x[:,k]+ k2/2)
        k4 = dt*f(t[k] + dt, x[:,k]+ k3)

        x[:,k+1] = x[:,k] + 1/6*(k1+2*k2+2*k3+k4)

    x=np.transpose(x)
    return x,t



# ==============================================================================
#  AB3 EXPLICIT SCHEME
def AdamsBashforth3(f, x0, t0, tf, dt):

    # Initializing variables
    t = np.arange(t0, tf, dt)
    nt = t.size
    nx = x0.size
    x=np.zeros((nt,nx))

    # Initial conditions and initizialization of AB3 method with Explicit Euler
    x[0,:] = x0
    x[1,:] = x[0,:] + dt * f(dt,x[0,:])
    x[2,:] = x[0,:] + dt * f(dt,x[1,:])

    for k in range(2,nt -1):
        x[k+1,:] = x[k,:] + dt/12*( 23* f(k*dt, x[k,:]) -16*f((k-1)*dt, x[k-1,:]) + 5*f((k-2)*dt, x[k-2,:]))
    
    return x,t



# ==============================================================================
#  POINCARE MAP PLOT
def PoincareMap(sol,t,omega,t_start):
    dt = t[1]-t[0]
    p_map=np.zeros((t.size,np.size(sol[0,:])))

    for k in range(len(t)):
        ind=ceil(t_start+1/dt*2*pi*k/omega)
        if ind <= len(t):
            p_map[k,:] = sol[ind,:]
        else:
            p_map[k,:] = []
            # continue
    
    # plt.plot(p_map[:,0],p_map[:,1],'bs',label="RK4")
    plt.figure(figsize=(10,6))
    title = r"$\ddot{x} +\delta \dot{x}+\alpha x +\beta x^3 = \gamma \cos(\omega t)$"
    plt.title(title)
    plt.xlim([-1.6, 1.6])
    plt.ylim([-1.25,1.25])
    plt.scatter(p_map[:,0],p_map[:,1],s=2, c="coral")
    plt.savefig('plot'+str(t_start)+'.png',dpi=300)
    plt.close()
    return



# ==============================================================================
#  NUMERICAL SOLUTION

#  Definition of Duffing Oscillator parameters and equivalent I order system
params = {"alpha": 1, "beta": 5, "gamma": 8, "delta":0.02, "omega":0.5}
f = lambda t,x : Duffing(t,x, params)

# Initial conditions and disretization
x0=np.array([0,0])
t0=0
tf=50
N=4000
dt=tf/N


# Solution by Runge-Kutta 4 - Explicit scheme
sol,t = RungeKutta4(f,x0,t0,tf,dt)
x_RK4 = sol[:,0]
xdot_RK4 = sol[:,1]


# Solution by Adams-Bashforth 3 - Explicit scheme
sol1,t = AdamsBashforth3(f,x0,t0,tf,dt)
x_AB3 = sol1[:,0]
xdot_AB3 = sol1[:,1]



# ==============================================================================
#  SEQUENCE OF POINCARE MAPS: FOR A COMPLETE ANIMATION SET
#   tf=10000   N=6000000, change constants to: 
#   params = {"alpha": -1, "beta": 1, "gamma": 0.3, "delta":0.15, "omega":1}
# and decomment the follow

# for k in range (1,100):
#     PoincareMap(sol1,t,params["omega"],1+3*k)



# ==============================================================================
#  PLOT
plt.figure(dpi=150)
plt.figure(figsize=(10,6))
title = 'Duffing Oscillator \n ' + r'$\ddot{x} +\delta \dot{x}+\alpha x +\beta x^3 = \gamma \cos(\omega t)$'
plt.title(title)
plt.plot(t,x_RK4, "r",label="RK4")
plt.plot(t,x_AB3, "b", label="AB3")
plt.xlabel("Time")
plt.ylabel("x")
plt.grid()
plt.legend()
plt.show()
